Add support for linux s390x gnu

edlerd opened the pull request:
This PR contains:

- [ ] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [x] other

Are tests included?

- [ ] yes (_bugfixes and features will not be merged without tests_)
- [x] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:

### Description

Add support for s390x linux gnu target.

I would very much appreciate it if a new release with s390x support could be published after we get this PR merged :)

pavolloffay commented:
We are interested in s390x support as well

pavolloffay commented:
@edlerd we are getting this error:

```
Your current platform "linux" and architecture "s390x" combination is not yet supported by the native Rollup build. Please use the WASM build "@rollup/wasm-node" instead.
```

Does the workaround with `rollup/wasm-node` work?

edlerd commented:
> Does the workaround with `rollup/wasm-node` work?

We need it in combination with vite 5.0 -- which uses rollup 4. It seems to work with `rollup/wasm-node` on s390x at first, but then `vite build` crashes in our build with the below trace

```
memory access out of bounds (Note that you need plugins to import files that are not JavaScript)
:: file: /build/lxd/parts/lxd-ui/build/index.html
:: error during build:
:: RuntimeError: memory access out of bounds
::     at wasm://wasm/0063bf96:wasm-function[1558]:0x162080
::     at wasm://wasm/0063bf96:wasm-function[249]:0x104dc0
::     at wasm://wasm/0063bf96:wasm-function[1442]:0x15f2f9
::     at module.exports.parse (/build/lxd/parts/lxd-ui/build/node_modules/rollup/dist/wasm-node/bindings_wasm.js:135:14)
::     at exports.parseAsync (/build/lxd/parts/lxd-ui/build/node_modules/rollup/dist/native.js:5:2)
::     at parseAstAsync (file:///build/lxd/parts/lxd-ui/build/node_modules/rollup/dist/es/shared/parseAst.js:2148:29)
::     at Module.tryParseAsync (file:///build/lxd/parts/lxd-ui/build/node_modules/rollup/dist/es/shared/node-entry.js:13517:26)
::     at Module.setSource (file:///build/lxd/parts/lxd-ui/build/node_modules/rollup/dist/es/shared/node-entry.js:13098:46)
::     at ModuleLoader.addModuleSource (file:///build/lxd/parts/lxd-ui/build/node_modules/rollup/dist/es/shared/node-entry.js:17788:26)
```

I opened this PR to unblock the build. Hints on how to fix the build are highly welcome :)

pavolloffay commented:
I have submitted a similar https://github.com/rollup/rollup/pull/5350 for ppc64le

lukastaegert reviewed (commented):
Please have a look at CI, unfortunately it does not work yet.

edlerd commented:
> Please have a look at CI, unfortunately it does not work yet.

Fixed the lint error, though no clue yet how to fix the [build error](https://github.com/rollup/rollup/actions/runs/7568759831/job/20663909080?pr=5346).

iblancasa commented:
I think this error message is related @edlerd:
```sh
 2024-01-19T19:12:21.892Z napi:build Run cargo build --release --target s390x-unknown-linux-gnu -p bindings_napi
/bin/sh: 1: zig: not found
2024-01-19T19:12:21.895Z napi:build Could not find zig on the PATH, fallback to normal linker
```

I cannot see that error message in the build for other architectures. `ld` seems to not be configured properly for cross-compilation... trying to link (I guess) x64 stuff with s390x stuff.

uweigand commented:
We just ran into this issue as well - thanks for working on s390x support, @edlerd!

> `ld` seems to not be configured properly for cross-compilation... trying to link (I guess) x64 stuff with s390x stuff.

Normally, cargo should be able to use the proper cross-linker.  You need to set the environment variable
```
CARGO_TARGET_S390X_UNKNOWN_LINUX_GNU_LINKER=s390x-linux-gnu-gcc
```

I'm not very familiar with the napi build script, but from a quick glance it seems it should already attempt to set this up, but this doesn't appear to be working correctly (you can see in the error message that cargo just falls back to `cc` as linker).   You might try just setting the above environment variable directly in the build-and-tests.yml rule.

edlerd commented:
> Normally, cargo should be able to use the proper cross-linker. You need to set the environment variable
> 
> ```
> CARGO_TARGET_S390X_UNKNOWN_LINUX_GNU_LINKER=s390x-linux-gnu-gcc
> ```

Thanks for the suggestion, I added this just before calling the build (also rebased on current master branch). HTH

uweigand commented on .github/workflows/build-and-tests.yml:
```diff
@@ -183,6 +183,17 @@ jobs:
+              sudo apt-get install gcc-s390x-linux-gnu -y
+            build: >-
+              set -e &&
+              CARGO_TARGET_S390X_UNKNOWN_LINUX_GNU_LINKER=s390x-linux-gnu-gcc &&
```
Don't you need an "export" here (given that it is in its own bash statement)?

edlerd replied:
Good catch, added it to the line.

uweigand commented:
I've now tried a cross-build x86_64 -> s390x using this PR and the linker environment variable, and this completed successfully for me.

lukastaegert commented:
This CI run actually looked quite good! Now the only missing thing is `s390x-unknown-linux-gnu-strip: command not found`. Of course we can live without stripping debug symbols for the initial effort, but they make the binary huge.

uweigand commented:
> This CI run actually looked quite good! Now the only missing thing is `s390x-unknown-linux-gnu-strip: command not found`. Of course we can live without stripping debug symbols for the initial effort, but they make the binary huge.

Ah, that should be `s390x-linux-gnu-strip` (without the unknown).   That should be present as part of the standard Ubuntu cross-toolchain.

edlerd commented:
> > This CI run actually looked quite good! Now the only missing thing is `s390x-unknown-linux-gnu-strip: command not found`. Of course we can live without stripping debug symbols for the initial effort, but they make the binary huge.
> 
> Ah, that should be `s390x-linux-gnu-strip` (without the unknown). That should be present as part of the standard Ubuntu cross-toolchain.

Thanks again for the hint, updated the PR accordingly.

lukastaegert reviewed (approved):
This looks amazing now! Let's hope it actually works 😉
And thank you so much for sticking with it, I will make sure we can release it maybe as early as today.

uweigand commented:
> This looks amazing now! Let's hope it actually works 😉 And thank you so much for sticking with it, I will make sure we can release it maybe as early as today.

Thank you very much for your support!
